File: 1674-minimum-moves-to-make-array-complementary/1674-minimum-moves-to-make-array-complementary.cpp
class Solution {
public:
    int minMoves(vector<int>& nums, int limit) {
        int n = nums.size();
        vector<int> diff(2 * limit + 2, 0);
        int a,b,low,high,sum;
        for(int i = 0; i < n / 2; i++) {
            a = nums[i];
            b = nums[n - 1 - i];
            low = min(a, b) + 1;
            high = max(a, b) + limit;
            sum = a + b;
            diff[2] += 2;
            diff[2 * limit + 1] -= 2;
            diff[low] -= 1;
            diff[high + 1] += 1;
            diff[sum] -= 1;
            diff[sum + 1] += 1;
        }
        int ans = INT_MAX;
        int moves = 0;
        for(int target = 2; target <= 2 * limit; target++) {
            moves += diff[target];
            ans = min(ans, moves);
        }
        return ans;
    }
};

File: 1674-minimum-moves-to-make-array-complementary/1674-minimum-moves-to-make-array-complementary.java
class Solution {
    public int minMoves(int[] nums, int limit) {
        int n = nums.length;
        int[] diff = new int[2 * limit + 2];
        for (int i = 0; i < n / 2; i++) {
            int a = nums[i];
            int b = nums[n - 1 - i];
            int low = Math.min(a, b) + 1;
            int high = Math.max(a, b) + limit;
            int sum = a + b;
            diff[2] += 2;
            diff[2 * limit + 1] -= 2;
            diff[low] -= 1;
            diff[high + 1] += 1;
            diff[sum] -= 1;
            diff[sum + 1] += 1;
        }
        int ans = Integer.MAX_VALUE;
        int moves = 0;
        for (int target = 2; target <= 2 * limit; target++) {
            moves += diff[target];
            ans = Math.min(ans, moves);
        }
        return ans;
    }
}